testnet/src/burnchains/bitcoin_regtest_controller.rs

```diff
@@ -0,0 +1,710 @@
+use stacks::deps::bitcoin::util::hash::Sha256dHash;
+use stacks::net::StacksMessageCodec;
+use stacks::util::hash::{Hash160, hex_bytes};
+use stacks::util::secp256k1::Secp256k1PublicKey;
+use stacks::util::sleep_ms;
+
+pub struct BitcoinRegtestController {
+    config: Config,
+    indexer_config: BitcoinIndexerConfig,
+    db: Option<BurnDB>,
+    chain_tip: Option<BurnchainTip>,
+}
+
+impl BitcoinRegtestController {
+
+    pub fn generic(config: Config) -> Box<dyn BurnchainController> {
+        Box::new(Self::new(config))
+    }
+
+    pub fn new(config: Config) -> Self {
+        
+        std::fs::create_dir_all(&config.node.get_burnchain_path())
+            .expect("Unable to create workdir");
+        
+        let res = SpvClient::init_block_headers(&config.burnchain.spv_headers_path, BitcoinNetworkType::Regtest);
+        if let Err(err) = res {
+            error!("Unable to init block headers: {}", err);
+            panic!()
+        }
+
+        let indexer_config = {
+            let burnchain_config = config.burnchain.clone();
+            BitcoinIndexerConfig {
+                peer_host: burnchain_config.peer_host,
+                peer_port: burnchain_config.peer_port,
+                rpc_port: burnchain_config.rpc_port,
+                rpc_ssl: burnchain_config.rpc_ssl,
+                username: burnchain_config.username,
+                password: burnchain_config.password,
+                timeout: burnchain_config.timeout,
+                spv_headers_path: burnchain_config.spv_headers_path,
+                first_block: burnchain_config.first_block,
+                magic_bytes: burnchain_config.magic_bytes
+            }
+        };
+                
+        Self {
+            config: config,
+            indexer_config,
+            db: None,
+            chain_tip: None,
+        }
+    }
+
+    fn setup_indexer_runtime(&mut self) -> (Burnchain, BitcoinIndexer) {
+        let network = "regtest".to_string();
+        let working_dir = self.config.get_burn_db_path();
+        let burnchain = match Burnchain::new(&working_dir,  &self.config.burnchain.chain, &network) {
+            Ok(burnchain) => burnchain,
+            Err(e) => {
+                error!("Failed to instantiate burnchain: {}", e);
+                panic!()    
+            }
+        };
+
+        let indexer_runtime = BitcoinIndexerRuntime::new(BitcoinNetworkType::Regtest);
+        let burnchain_indexer = BitcoinIndexer {
+            config: self.indexer_config.clone(),
+            runtime: indexer_runtime
+        };
+        (burnchain, burnchain_indexer)
+    }
+
+    fn receive_blocks(&mut self) -> BurnchainTip {
+        let (mut burnchain, mut burnchain_indexer) = self.setup_indexer_runtime();
+
+        let (block_snapshot, state_transition) = match burnchain.sync_with_indexer(&mut burnchain_indexer) {
+            Ok(res) => res,
+            Err(e) => {
+                error!("Unable to sync burnchain: {}", e);
```

Comment:
In testing this, I've gotten this to panic on the first block.  My fix in `feature/neon-network-and-fixes` was to create a first-block `BurnchainTip`.

Reply:
Thank you for the patch @jcnelson! Addressed with https://github.com/blockstack/stacks-blockchain/pull/1381/commits/ef4af9c3f3afb661b0d1d7bf2ecdd7333e56120d